src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -706,6 +747,143 @@ static void getFileReaderAndSchema(
+    {
+        auto target_copy = target_;
+        if (ignore_case)
+            boost::to_lower(target_copy);
+
+        for (size_t i = 0; i < type_->getSubtypeCount(); ++i)
+        {
+            auto field_name = type_->getFieldName(i);
+            if (ignore_case)
+                boost::to_lower(field_name);
+
+            if (startsWith(target_copy, field_name) && (target_copy.size() == field_name.size() || target_copy[field_name.size()] == '.'))
+            {
+                return {target_copy.size() == field_name.size() ? "" : target_.substr(field_name.size() + 1), type_->getSubtype(i)};
+            }
+        }
+        return {"", nullptr};
+    };
+
+    if (orc::STRUCT == orc_type->getKind())
+    {
+        auto next_type_and_target = search_struct_field(target, orc_type);
+        const auto & next_target = next_type_and_target.first;
+        const auto * next_orc_type = next_type_and_target.second;
+        return next_orc_type ? traverseDownORCTypeByName(next_target, next_orc_type, type, ignore_case) : nullptr;
+    }
+    else if (orc::LIST == orc_type->getKind())
+    {
+        /// For cases in which header contains subcolumns flattened from nested columns.
+        /// For example, "a Nested(x String, y Int64)" is flattened to "a.x Array(String), a.y Array(Int64)", and orc file schema is still "a array<struct<x string, y long>>".
+        /// In this case, we should skip possible array type and traverse down to its nested struct type.
+        const auto * array_type = typeid_cast<const DataTypeArray *>(removeNullable(type).get());
+        const auto * orc_nested_type = orc_type->getSubtype(0);
+        if (array_type && orc::STRUCT == orc_nested_type->getKind())
+        {
+            auto next_type_and_target = search_struct_field(target, orc_nested_type);
+            const auto & next_target = next_type_and_target.first;
+            const auto * next_orc_type = next_type_and_target.second;
+            if (next_orc_type)
+            {
+                /// Adjust CH type to avoid inconsistency between CH and ORC type brought by flattened Nested type.
+                type = array_type->getNestedType();
+                return traverseDownORCTypeByName(next_target, next_orc_type, type, ignore_case);
+            }
+            else
+                return nullptr;
+        }
+        else
+            return nullptr;
+    }
+    else
+        return nullptr;
+}
+
+static void updateIncludeTypeIds(
+    DataTypePtr type, const orc::Type * orc_type, bool ignore_case, std::unordered_set<UInt64> & include_typeids)
+{
+    // std::cout << "ch type:" << type->getName() << ", orc_type:" << orc_type->toString() << std::endl;
+
+    /// For primitive types, directly append column id into result
+    if (orc_type->getSubtypeCount() == 0)
+    {
+        include_typeids.insert(orc_type->getColumnId());
+        return;
+    }
+
+    auto non_nullable_type = removeNullable(type);
+    switch (orc_type->getKind())
```

Comment:
So, for complex types to read a value, we should include all children indexes? Index of a parent is not enough?

Reply:
Please, comment here